server/src/main/java/org/apache/druid/metadata/PhysicalDatasourceMetadataBuilder.java

```diff
@@ -0,0 +1,221 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.metadata;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.query.GlobalTableDataSource;
+import org.apache.druid.query.TableDataSource;
+import org.apache.druid.segment.column.ColumnType;
+import org.apache.druid.segment.column.RowSignature;
+import org.apache.druid.segment.column.Types;
+import org.apache.druid.segment.join.JoinableFactory;
+import org.apache.druid.server.SegmentManager;
+import org.apache.druid.timeline.SegmentSchema;
+
+import javax.annotation.Nullable;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ *
+ */
+public class PhysicalDatasourceMetadataBuilder
+{
+
+  private final SegmentManager segmentManager;
+  private final JoinableFactory joinableFactory;
+  private final ColumnTypeMergePolicy columnTypeMergePolicy;
+
+  @Inject
+  public PhysicalDatasourceMetadataBuilder(SegmentManager segmentManager, JoinableFactory joinableFactory)
+  {
+    this.segmentManager = segmentManager;
+    this.joinableFactory = joinableFactory;
+    this.columnTypeMergePolicy = new LeastRestrictiveTypeMergePolicy(); // todo get from config SegmentMetadataCacheConfig.getColumnTypeMergePolicy
+  }
+
+  @VisibleForTesting
+  @Nullable
+  // todo move DatasourceTable to server package
+  public PhysicalDatasourceMetadata buildDruidTable(String dataSource, final Set<SegmentSchema> schemaSet)
+  {
+    // Preserve order.
+    final Map<String, ColumnType> columnTypes = new LinkedHashMap<>();
+
+    for (SegmentSchema segmentSchema : schemaSet) {
+      final RowSignature rowSignature = segmentSchema.getRowSignature();
+      if (rowSignature != null) {
+        for (String column : rowSignature.getColumnNames()) {
+          final ColumnType columnType =
+              rowSignature.getColumnType(column)
+                          .orElseThrow(() -> new ISE("Encountered null type for column [%s]", column));
+
+          columnTypes.compute(column, (c, existingType) -> columnTypeMergePolicy.merge(existingType, columnType));
```

Comment:
I am not sure if we can use any merge policy here. are the segments ordered in any way?

Reply:
This logic was originally being applied on the segments in `SegmentMetadataCache`. The order of the segments was newest first. I will update to order the segments in the same way.

Reply:
Also, the merge policy is config driven, set in `SegmentMetadataCacheConfig`.